src/main/java/space/obminyashka/items_exchange/service/impl/LocationServiceImpl.java

```diff
@@ -130,24 +130,26 @@ public String createParsedLocationsFile(List<RawLocation> creatingData) throws I
 
     @Override
     public String createParsedLocsFile(List<LocationRaw> creatingData) throws IOException {
+        creatingData.sort(Comparator.comparing(LocationRaw::getAreaEn));
+
         Set<Area> areas = new HashSet<>();
         Set<District> districts = new HashSet<>();
         Set<City> cities = new HashSet<>();
 
         for (LocationRaw location : creatingData) {
-            if (!location.getAreaUa().isBlank() && !location.getAreaEn().isBlank()) {
-                Area area = new Area(location);
-                areas.add(area);
-
-                if (!location.getDistrictUa().isBlank() && !location.getDistrictEn().isBlank()) {
-                    District district = new District(location, area);
-                    districts.add(district);
-
-                    if (!location.getCityUa().isBlank() && !location.getCityEn().isBlank()) {
-                        cities.add(new City(location, district));
-                    }
-                }
-            }
+            Area area = new Area(location);
+            List<LocationRaw> locationsWithDistrictInSelectedArea = creatingData.stream()
+                    .sorted(Comparator.comparing(LocationRaw::getDistrictEn))
+                    .filter(locationRaw -> locationRaw.getAreaEn().equals(area.getNameEn()))
+                    .toList();
+
+            locationsWithDistrictInSelectedArea.stream()
+                    .map(locationRaw -> new District(locationRaw, area))
+                    .forEach(district -> {
+                        districts.add(district);
+                        cities.addAll(getCities(locationsWithDistrictInSelectedArea, district));
+                    });
+            areas.add(area);
```

Comment:
Any particular reason why that cannot be used as a stream till the end using .collect(Collectors.toSet()) if you're doing that with a single passage?